### PR TITLE
[cli] Add dependency exclusion to `npx expo install`

### PR DESCRIPTION
**Pull request by keith-kurak**
# Why
Fulfills https://linear.app/expo/issue/ENG-8564/add-dependency-exclusion-to-npx-expo-install. Developers will be able to suppress warnings from `npx expo install --check` via a package.json config, and thus suppress doctor warnings when they intentionally choose to install a different version, especially when using dev builds and not needing to stick to the Expo Go- bundled version.

# How
`npx expo install --check|fix` reads from `expo.install.exclude` if it's available, and, if one of the packages it thinks it should fix is in that list, it will skip asking/ fixing it. If you run `EXPO_DEBUG=1 npx expo install --check`, the packages whose checks were actually skipped will be listed in debug output; otherwise there are no warnings or messages about the excluded packages.

The package.json configuration verbiage is intended to be similar to the [autolinking config](https://docs.expo.dev/modules/autolinking/#exclude).

~~It also seemed appropriate that `npx expo install` should behave differently if you try to install an excluded package, so someone doesn't come along and write over a version that another developer determined was good and thus applied the override for. Currently, it aborts if any of the packages are in the exclusion list. I could see doing a warning and skipping the affected packages instead, but the command outright failing is pretty hard to miss.~~

`npx expo install` now installs excluded packages with the latest version, just like a package without a specified native version. It will add a note describing what it's doing and why.

# Test Plan
- [x] test without `expo.install.exclude` (works the same)
- [x] test with empty exclusion list (works the same)
- [x] test with exclusions that aren't actual dependencies (works the same)
- [x] test `expo install --fix|check` when excluding an actual dependency (skips it)
- [x] test with `EXPO_DEBUG=1` (lists excluded packages)
- [x] test `expo install [package]` with exclusions (installs latest with notes)
- [x] test `expo install [package]` without exclusions (proceeds as it normally does)

## Output
`npx expo install --check`, showing note that it is skipping checking packages in the exclude list:
image
Additional debug output for `EXPO_DEBUG=1 npx expo install --check`:
image
`expo install <package>` with exclusions:
image

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Review by brentvatne — COMMENTED** [review]
looks good, let's also add some documentation in this same PR. we can add a note along with it that this field is only supported in sdk 49+

File: docs/pages/more/expo-cli.mdx [hunk 1]
```diff
@@ -315,6 +315,20 @@ The command `npx expo install expo-camera` and `npx expo install expo-camera --f
 
 <Terminal cmd={['$ npx expo install --fix']} />
 
+### Configuration
```
Comment (brentvatne) [inline docs/pages/more/expo-cli.mdx:318]:
```suggestion
### Configuring dependency validation

> Available in SDK 49 and higher.
```

File: docs/pages/more/expo-cli.mdx [hunk 2]
```diff
@@ -315,6 +315,20 @@ The command `npx expo install expo-camera` and `npx expo install expo-camera --f
...
+There may be circumstances where you need to use a version of a package that is outside the range of safe versions recognized by `npx expo install`. In SDK 49 and higher projects, you can exclude specific packages from the version checks, while still allowing the `install` command to install, check, and fix any other dependencies. This configuration extends to the checking done by `npx expo-doctor`, suppressing warnings related to these specific packages during the Expo Doctor step in EAS Build.
```
Comment (brentvatne) [inline docs/pages/more/expo-cli.mdx:319]:
```suggestion
There may be circumstances where you want to use a version of a package is different from the version recommended by `npx expo install`. For example, you might test a new version of `react-native-reanimated`, verify that it works well in your app and fixes a bug that you encountered, and now you want to 1) not be warned by `npx expo start` or `npx expo-doctor` and 2) not have that package version changed when you run `npx expo install --fix`. 

In SDK 49 and higher projects, you can exclude specific packages from the version checks, while still allowing the `install` command to install, check, and fix any other dependencies. This configuration extends to the checking done by `npx expo-doctor`.
```

**Review by brentvatne — CHANGES_REQUESTED** [review]

**Comment by brentvatne** [comment]
another suggestion - maybe we should link to the docs in the validation message? so if someone has an "invalid" dep version and want to have that one be ignored, the information on how they do that is available directly in the message.

File: docs/pages/more/expo-cli.mdx (hunk 1 above)
Comment (amandeepmittal) [reply docs/pages/more/expo-cli.mdx:318]: One small nit:

> Available in SDK 49 and above.

File: docs/pages/more/expo-cli.mdx (hunk 2 above)
Comment (amandeepmittal) [reply docs/pages/more/expo-cli.mdx:319]: My suggestion on phrasing the example part with a line break (not having line break, + longer paragraphs, are hard to skim and understand at the same time:

```markdown
For example, you want to test a new version of `react-native-reanimated` to verify that it works well in your app and fixes a bug that you encountered. Now you want 1) not be warned by `npx expo start` or `npx expo-doctor` and 2) not have that package version changed when you run `npx expo install --fix`.
```

I'd also suggest to remove the `,` after checks (not grammatically correct) and no need for pause here:

```markdown
...specific packages from the version checks while still...
```

File: docs/pages/more/expo-cli.mdx [hunk 3]
```diff
@@ -315,6 +315,26 @@ The command `npx expo install expo-camera` and `npx expo install expo-camera --f
 
 <Terminal cmd={['$ npx expo install --fix']} />
 
+### Configuring dependency validation
+
+> Available in SDK 49 and above.
+There may be circumstances where you want to use a version of a package is different from the version recommended by `npx expo install`.
```
Comment (amandeepmittal) [inline docs/pages/more/expo-cli.mdx:321]: @keith-kurak, is this line supposed to be part of the callout? 

CleanShot 2023-06-11 at 23 33 02@2x

Comment (keith-kurak) [reply docs/pages/more/expo-cli.mdx:322]: Nope, didn't mean that! Added a line break to fix

**Review by amandeepmittal — APPROVED** [review]
Thanks for fixing the callout issue!

**Review by brentvatne — APPROVED** [review]

**Pull request merged**